### PR TITLE
 Add support for setting --history-max on existing clusters

### LINKED ISSUE #3414: Set --history-max on a already deployed cluster
for the sake of cleanlyness, I'm really interested on adding this option to my cluster.

Reading various thread didn't lead me to a clear path.
I tried various things without success. Would be glad to get some help.

```
helm init --upgrade --history-max=10
```

I think this did work, but I'm still seeing my old ConfigMaps (even after an upgrade of the chart)

How can I confirm that Tiller is configured now. And how can I delete old revisions?
(running v2.8.0)

Thanks!

### PR DESCRIPTION
**Pull request by rossf7**
**What this PR does / why we need it**:

This PR sets `--history-max` on existing clusters. It is a replacement for #3454 and includes the unit tests added in af316d3. I've tested locally and is working OK.

closes #3414

**Special notes for your reviewer**:

I changed the unit tests in af316d3 quite a lot to try and sync them with the other tests. The code change is basically unchanged from the original PR.

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

File: cmd/helm/installer/install.go [hunk 1]
```diff
@@ -71,6 +72,11 @@ func Upgrade(client kubernetes.Interface, opts *Options) error {
 	}
 	obj.Spec.Template.Spec.Containers[0].Image = opts.SelectImage()
 	obj.Spec.Template.Spec.Containers[0].ImagePullPolicy = opts.pullPolicy()
+	if opts.MaxHistory > 0 {
```
Comment (bacongobbler) [inline cmd/helm/installer/install.go]: how would a user unset this value?

Comment (rossf7) [reply cmd/helm/installer/install.go]: Good point. If `opts.MaxHistory` is not set should we revert to the default of `0`?

@bacongobbler I'd appreciate hearing what you think is the best approach? Thanks!

File: cmd/helm/installer/install.go [hunk 2]
```diff
@@ -407,3 +423,32 @@ func generateSecret(opts *Options) (*v1.Secret, error) {
 }
 
 func read(path string) (b []byte, err error) { return ioutil.ReadFile(path) }
+
+// updateEnv updates the existing env vars and allows env vars to be removed
+// by name.
+func updateEnv(existing []v1.EnvVar, env []v1.EnvVar, remove []string) []v1.EnvVar {
```
Comment (rossf7) [inline cmd/helm/installer/install.go]: Also, this came from the previous PR and I have a doubt now. Do we need to be able to remove env vars?

As doesn't `TILLER_HISTORY_MAX` always need to be set.

**Comment by alexcurtin** [comment]
Any update on this?

**Comment by bacongobbler** [comment]
Hey @alexcurtin, thanks for the ping.

Unfortunately, we won't be able to get this feature into 2.15, which will be the last feature release for Helm 2. Because Tiller is no longer present in Helm 3, it is hard to justify major changes (which could incur support overhead) to that subsystem.

The good news is that [workarounds exist](https://github.com/helm/helm/issues/3414#issuecomment-361497289), and because Tiller is not present in Helm 3, when you've upgraded you should no longer need this workaround.

Thanks!

**Pull request closed without merging**